Ktisis/History/HistoryManager.cs

```diff
@@ -0,0 +1,282 @@
+                if (!isBoneInHistory) AddEntryToHistory(matrix, bone);
+            }
+
+            if ((newState == TransformTableState.IDLE) && (_currentTtState == TransformTableState.EDITING))
+            {
+                PluginLog.Information("Finished TT edit");
+                AddEntryToHistory(matrix, bone);
+            }
+
+            _currentTtState = newState;
+        }
+
+        private unsafe void AddEntryToHistory(Matrix4x4 tt, Bone? bone)
+        {
+            History!.Insert(_maxIdx, new(tt, bone, (Actor*)Ktisis.GPoseTarget!.Address));
+            _currentIdx++;
+            _maxIdx++;
+            PluginLog.Information($"Current Idx: {_currentIdx} - Max Idx: {_maxIdx}");
+            printHistory(_maxIdx);
+        }
+
+        private void printHistory(int until)
+        {
+            var str = "\n";
+            for (int i = 0; i < until; i++)
+            {
+                str += $"{i + 1}: ";
+                var entry = History![i];
+                if (entry.Bone is null) str += $"Bone Global";
+                else str += $"Bone {Locale.GetBoneName(entry.Bone!.HkaBone.Name.String)}";
+                str += "\n";
+            }
+            PluginLog.Information(str);
+        }
+
+        private unsafe void OnGizmoChange(GizmoState state)
+        {
+            if (!PoseHooks.PosingEnabled) return;
+
+            var newState = state;
+            var bone = Skeleton.GetSelectedBone(EditActor.Target->Model->Skeleton);
+            var boneTransform = bone!.AccessModelSpace(PropagateOrNot.DontPropagate);
+            var matrix = Interop.Alloc.GetMatrix(boneTransform);
+
+            if ((newState == GizmoState.EDITING) && (_currentGizmoState == GizmoState.IDLE))
+            {
+                PluginLog.Information("Started Gizmo edit");
+                if (_maxIdx != _currentIdx) alternativeTimelineWarning();
+                var isBoneInHistory = History?.FirstOrDefault(historyItem => historyItem.Bone?.UniqueName == bone?.UniqueName) != null;
+                if (!isBoneInHistory) AddEntryToHistory(matrix, bone);
+            }
+            if ((newState == GizmoState.IDLE) && (_currentGizmoState == GizmoState.EDITING))
+            {
+                PluginLog.Information("Ended Gizmo edit");
+                AddEntryToHistory(matrix, bone);
+            }
+            _currentGizmoState = newState;
+        }
+
+
+        private static HistoryManager? _instance;
+
+        public static HistoryManager Instance
+        {
+            get
+            {
+                if (_instance == null)
+                {
+                    _instance = new HistoryManager();
+                }
+                return _instance!;
+            }
+        }
+
+        /* 
+         This is weird yes. 
+         But it exists to make it obvious that it needs to be created first so it can subscribe to the Framework.Update event.
+        */
+        public static void Init() 
+        {
+            _ = Instance;
```

Comment:
This initialization only works once per time the assembly is loaded, so disabling and reenabling the plugin would prevent this class from listening to the events (to my understanding)

Reply:
After disabling and reenabling the plugin, HistoryManager is still listening to OnTransformationMatrixChange and OnGizmoChange. I don't know if that's probably because I have it as a dev plugin or not.